src/mscorlib/shared/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs

```diff
@@ -203,6 +224,112 @@ public override void Return(T[] array, bool clearArray = false)
+            }
+
+            ArrayPoolEventSource log = ArrayPoolEventSource.Log;
+            uint trimTicks = pressure == MemoryPressure.High ? TlsHighTrimAfterTicks : TlsTrimAfterTicks;
+            foreach (var tlsBuckets in s_allTlsBuckets)
+            {
+                Bucket[] buckets = tlsBuckets.Key;
+                for (int i = 0; i < NumBuckets; i++)
+                {
+                    uint stockTicks = buckets[i].Ticks;
+                    if (stockTicks > tickCount || (tickCount - stockTicks) > trimTicks)
+                    {
+                        T[] buffer = buckets[i].Buffer;
+
+                        // We've wrapped the tick count or elapsed enough time since the
+                        // first item went into this thread local bucket. Clear it.
+                        buckets[i].Buffer = null;
+
+                        if (log.IsEnabled() && buffer != null)
+                        {
+                            log.BufferTrimmed(buffer.GetHashCode(), buffer.Length, Id);
+                        }
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// This is the static function that is called from the gen2 GC callback.
+        /// The input object is the instance we want the callback on.
+        /// </summary>
+        /// <remarks>
+        /// The reason that we make this function static and take the instance as a parameter is that
+        /// we would otherwise root the instance to the Gen2GcCallback object, leaking the instance even when
+        /// the application no longer needs it.
+        /// </remarks>
+        private static bool Gen2GcCallbackFunc(object target)
+        {
+            return ((TlsOverPerCoreLockedStacksArrayPool<T>)(target)).Trim();
+        }
+
+        private enum MemoryPressure
+        {
+            Low,
+            Medium,
+            High
+        }
+
+        private static MemoryPressure GetMemoryPressure()
+        {
+            GC.GetMemoryInfo(out uint threshold, out _, out uint lastLoad, out _, out _);
+            if (lastLoad >= threshold * .90)
+            {
+                return MemoryPressure.High;
+            }
+            else if (lastLoad >= threshold * .80)
+            {
+                return MemoryPressure.Medium;
+            }
+            return MemoryPressure.Low;
+        }
+
+        private static bool TrimBuffers { get; } = GetTrimBuffers();
+
+        private static bool GetTrimBuffers()
+        {
+            // Environment uses ArrayPool, so we have to hit the API directly.
+#if !CORECLR
+            // P/Invokes are different for CoreCLR/RT- for RT we'll not allow
+            // enabling/disabling for now.
+            return true;
+#else
+            Span<char> buffer = stackalloc char[32];
+            int result = Win32Native.GetEnvironmentVariable(@"DOTNET_SYSTEM_BUFFERS_TRIMBUFFERS", buffer);
```

Comment:
I wonder about the value of this environment variable at all.    We should seriously consider removing it, as its only value right now is a 'perf - compat' switch and we do far more dangerous changes than this without such a fallback.   I see this is 'cruft' that we will end up supporting forever for no good reason.  

 What MIGHT be useful is a variable that indicates (e.g. on a scale form 0-100) how aggressive we should be about cleaning up caches.   However this should be some GLOBAL setting about how precious memory is (we wanted it for the GC already to decide how aggressive it should be about collecting), and this code is just 'following suit'.

Thus I would recommend removing this code.   At the very least, get it out of the startup (critical) path.     We assume this stuff will be on 99% of the time, so we should only check this variable during Trim.

Reply:
As it allows not creating the callback I don't know that we can get rid of it without more impact than not. The code is reasonably performant (doesn't allocate, or do much work).

 From my discussions with @jkotas we do believe the kill/enable switch is important in the short term.

As far as more granular tuning- we are listening to the GC. In theory it is possible to tweak the threshold there and we'll be impacted here. More than that I think we need to discuss carefully.

Reply:
> we do believe the kill/enable switch is important in the short term.

I meant it just as mitigation for pushing this change through the system aggressively with a short runway (to ensure that people do not get blocked if this does not work as expected). I do not have a strong opinion about this.

Reply:
I believe we should, for the reasons you mention.